### PR TITLE
remove staker property in the delegation pools

### PR DESCRIPTION
**Pull request by Vitomir2**
- delete the staker from the DelegationPool lib;
- comment the NoTokensDelegated error and its revert, and add clarification comments;
- exit the pool reward distribution if the virtualSupply = 0 to avoid Panic;
- create a new test to cover the case of virtualSupply = 0 and add a new virtualSupply getter;
- adapt the contracts and the rest forge tests;

### CODE REVIEW

File: contracts/HydraDelegation/HydraDelegation.sol [hunk 1]
```diff
@@ -87,6 +87,13 @@ contract HydraDelegation is
         _distributeDelegationRewards(staker, reward, epochId);
     }
 
+    /**
+     * @inheritdoc IHydraDelegation
+     */
+    function setDelegationPoolStaker(address staker) external onlyHydraChain {
```
Comment (R-Santev) [inline contracts/HydraDelegation/HydraDelegation.sol:93]: "initDelegationPool" sounds better to me

**Review by R-Santev — CHANGES_REQUESTED** [review]
I saw that the only place "staker" param is used is as argument of custom error. We would be able to easily understand who is the staker in case the custom error is triggered without passing the staker as a property, so it is better to just remove the parameter of the custom error and then remove the staker from the DelegationPool struct, because we keep the staker as the key of the mapping

File: contracts/HydraDelegation/HydraDelegation.sol (hunk 1 above)
Comment (Vitomir2) [reply contracts/HydraDelegation/HydraDelegation.sol:93]: Yeah, I actually thought about this, but we are actually initializing just the staker property. However, will rename it because I like it more, too.

File: contracts/HydraDelegation/DelegationPoolLib.sol [hunk 2]
```diff
@@ -21,9 +22,10 @@ library DelegationPoolLib {
      * @param pool the DelegationPool for rewards to be distributed to
      * @param amount the total amount to be distributed
      */
-    function distributeReward(DelegationPool storage pool, address staker, uint256 amount) internal {
+    function distributeReward(DelegationPool storage pool, uint256 amount) internal {
         if (amount == 0) return;
-        if (pool.virtualSupply == 0) revert NoTokensDelegated(staker);
+        // In our version of the contracts we cannot hit this case, because virtualSupply will always equal the supply
```
Comment (R-Santev) [inline contracts/HydraDelegation/DelegationPoolLib.sol:27]: And what if supply is zero?

File: test/forge/libs/DelegationPool.t.sol [hunk 3]
```diff
@@ -57,17 +57,18 @@ contract DelegationPoolTest is Test {
...
         // did not write to storage
         (, bytes32[] memory writes) = (vm.accesses(address(this)));
         assertEq(writes.length, 0);
     }
 
-    function testCannotDistributeReward_NoTokensDelegated() public {
-        vm.expectRevert(abi.encodeWithSelector(NoTokensDelegated.selector, (address(delegationPoolLibUser))));
-        delegationPoolLibUser.distributeReward(address(delegationPoolLibUser), 1);
-    }
+    // In our version of the contracts we cannot hit this case
```
Comment (R-Santev) [inline test/forge/libs/DelegationPool.t.sol:67]: Add a test that tries to distribute rewards where the delegator pool is empty, so if something is changed in the wrong way we can see the test is failing

**Review by R-Santev — CHANGES_REQUESTED** [review]

File: contracts/HydraDelegation/DelegationPoolLib.sol (hunk 2 above)
Comment (Vitomir2) [reply contracts/HydraDelegation/DelegationPoolLib.sol:27]: It should just leave the function, otherwise the calculation will fail

**Review by R-Santev — APPROVED** [review]
